compiler/package.go

```diff
@@ -468,39 +468,38 @@ func Compile(importPath string, files []*ast.File, fileSet *token.FileSet, impor
 				c.Printf(`%s = $newType(%d, %s, "%s.%s", %t, "%s", %t, %s);`, lhs, size, typeKind(o.Type()), o.Pkg().Name(), o.Name(), o.Name() != "", o.Pkg().Path(), o.Exported(), constructor)
 			})
 			d.MethodListCode = c.CatchOutput(0, func() {
-				if _, isInterface := o.Type().Underlying().(*types.Interface); isInterface {
+				if o.IsAlias() {
```

Comment:
Shouldn't this filter be already at a higher level, e.g. at line 436? As far as I can see, type aliases only exist at the source level, they should not occur in the compiler output any more.

Reply:
That's my understanding too.

Reply:
Quite possibly. However, I can't answer with confidence, it would require further investigation.

I am quite confident in its current location because of my understanding of methods (a type alias may not have any additional methods; they're always associated with the aliased type instead), and also because I've inspected the generated code and confirmed it looked good.

One thing to find out, if this is moved higher level, is what happens regarding DCE when a type is only ever referenced via its type alias. E.g.:

```Go
package p

type T int

type A = T
```

```Go
package main

import "p"

func main() {
    var a p.A = 123
    fmt.Printf("%T %v\n", a, a)
}
```

Reply:
I wouldn't anticipate there being a problem in this situation, because everything is done based on the derived type of expressions. Aliases have no impact on that calculation because they've effectively already been erased.

Reply:
e.g.

```go
package main

import (
	"fmt"
	"go/ast"
	"go/importer"
	"go/parser"
	"go/token"
	"go/types"
	"log"
)

const hello = `package main

import "bytes"

type T = bytes.Buffer

var x T

func main() {
}
`

func main() {
	fset := token.NewFileSet()

	f, err := parser.ParseFile(fset, "hello.go", hello, 0)
	if err != nil {
		log.Fatal(err)
	}

	conf := types.Config{Importer: importer.Default()}

	pkg, err := conf.Check("cmd/hello", fset, []*ast.File{f}, nil)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("x: %s\n", pkg.Scope().Lookup("x").Type())
}
```

gives:

```
x: bytes.Buffer
```

hence the dependency on `bytes` is derived.

Reply:
Fixed typo in my closing sentence...

Reply:
That sounds reasonable. I don't have objections to trying it out. The basic tests I've tried all pass.

Reply:
Cool! I prefer this solution because it is less likely to hide bugs (e.g. some bad runtime reference to a type alias).